src/python/pants/goal/migrate_call_by_name.py

```diff
@@ -3,35 +3,655 @@
+
+        try:
+            if len(get.args) == 1:
+                new_source, imports = self.map_no_args_get_to_new_syntax(get, get_deps)
+            elif len(get.args) == 2 and isinstance(get.args[1], ast.Call):
+                new_source, imports = self.map_short_form_get_to_new_syntax(get, get_deps)
+            elif len(get.args) == 2 and isinstance(get.args[1], ast.Dict):
+                new_source, imports = self.map_dict_form_get_to_new_syntax(get, get_deps)
+            elif len(get.args) == 3:
+                new_source, imports = self.map_long_form_get_to_new_syntax(get, get_deps)
+            else:
+                logging.warning(f"Failed to migrate {ast.unparse(get)} due to unknown form\n")
+                return None
+        except AssertionError as e:
+            logging.warning(f"Failed to migrate {ast.unparse(get)} with assertion error {e}\n")
+            return None
+        except Exception as e:
+            logging.warning(f"Failed to migrate {ast.unparse(get)} with {e}\n")
+            return None
+
+        assert get.end_lineno is not None
+        assert get.end_col_offset is not None
+        return Replacement(
+            filename=filename,
+            module=graph_item["module"],
+            line_range=(get.lineno, get.end_lineno),
+            col_range=(get.col_offset, get.end_col_offset),
+            current_source=get,
+            new_source=new_source,
+            additional_imports=[
+                ast.ImportFrom(
+                    module="pants.engine.rules", names=[ast.alias("implicitly")], level=0
+                ),
+                *imports,
+            ],
+        )
+
+    def map_no_args_get_to_new_syntax(
+        self, get: ast.Call, deps: list[RuleGraphGetDep]
+    ) -> tuple[ast.Call, list[ast.ImportFrom]]:
+        """Get(<OutputType>) -> the_rule_to_call(**implicitly())"""
+
+        logger.debug(ast.dump(get, indent=2))
+        assert len(get.args) == 1, f"Expected 1 arg, got {len(get.args)}"
+        assert isinstance(output_type := get.args[0], ast.Name), f"Expected Name, got {get.args[0]}"
+
+        dep = next(
+            dep
+            for dep in deps
+            if dep["output_type"]["name"] == output_type.id and not dep["input_types"]
+        )
+        module = dep["rule_dep"]["module"]
+        new_function = dep["rule_dep"]["function"]
+
+        new_call = ast.Call(
+            func=ast.Name(id=new_function),
+            args=[],
+            keywords=[
+                ast.keyword(value=ast.Call(func=ast.Name(id="implicitly"), args=[], keywords=[]))
+            ],
+        )
+        imports = [ast.ImportFrom(module, names=[ast.alias(new_function)], level=0)]
+        return new_call, imports
+
+    def map_long_form_get_to_new_syntax(
+        self, get: ast.Call, deps: list[RuleGraphGetDep]
+    ) -> tuple[ast.Call, list[ast.ImportFrom]]:
+        """Get(<OutputType>, <InputType>, input) -> the_rule_to_call(**implicitly(input))"""
```

Comment:
I'm not seeing why there is a difference between `map_long_form_get_to_new_syntax` and `map_short_form_get_to_new_syntax` in terms of the _output_ syntax: it's clear why the _input_ is different, but the "if the called function only takes a single argument with a type that matches the argument" case applies to both of them I think.

Reply:
Yeah, I updated the short-form syntax before pushing this PR, but didn't give much thought to the long-form at the time. 

From the examples I've seen in the code, the "short-form" syntax calls the XYZRequest constructor, so replacing that is pretty simple. 

But the long-form calls seem to have a wider variety of call patterns, and one of the original calls I tried to replace similar to the short-form ended up failing the type checker due to some missing context (but works when called using `implicitly(dict)`